cudax/include/cuda/experimental/__device/device_vector.cuh

```diff
@@ -0,0 +1,231 @@
+
+  _CCCL_NODISCARD _CUDA_VSTD::size_t size() const;
+
+  _CCCL_NODISCARD iterator begin() const noexcept;
+
+  _CCCL_NODISCARD iterator end() const noexcept;
+};
+
+//! @brief A random-access iterator over all available CUDA devices
+struct device_vector::iterator
+{
+  using value_type        = device;
+  using difference_type   = int;
+  using pointer           = const device*;
+  using reference         = device;
+  using iterator_category = ::cuda::std::random_access_iterator_tag;
+  using iterator_concept  = ::cuda::std::random_access_iterator_tag;
+
+  //! @brief Proxy object used as the return value of `operator->`
+  struct __proxy
+  {
+    _CCCL_NODISCARD const device* operator->() const noexcept
+    {
+      return &__dev_;
+    }
+
+    device __dev_;
+  };
+
+  _CCCL_NODISCARD constexpr device operator*() const noexcept
+  {
+    return device(__id_);
+  }
+
+  _CCCL_NODISCARD constexpr device operator[](int __offset) const noexcept
+  {
+    return device(__id_ + __offset);
+  }
+
+  _CCCL_NODISCARD constexpr __proxy operator->() const noexcept
+  {
+    return __proxy{device(__id_)};
+  }
+
+  constexpr iterator& operator++() noexcept
+  {
+    ++__id_;
+    return *this;
+  }
+
+  constexpr iterator operator++(int) noexcept
+  {
+    auto tmp = *this;
+    ++__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator--() noexcept
+  {
+    --__id_;
+    return *this;
+  }
+
+  constexpr iterator operator--(int) noexcept
+  {
+    auto tmp = *this;
+    --__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator+=(int __offset) noexcept
+  {
+    __id_ += __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator+(int __offset) noexcept
+  {
+    return iterator(__id_ + __offset);
+  }
+
+  constexpr iterator& operator-=(int __offset) noexcept
+  {
+    __id_ -= __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator-(int __offset) noexcept
+  {
+    return iterator(__id_ - __offset);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr iterator operator+(int __lhs, const iterator& __rhs) noexcept
+  {
+    return iterator(__lhs + __rhs.__id_);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr int operator-(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ - __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator==(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ == __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator!=(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ != __rhs.__id_;
+  }
+
+private:
+  friend struct device_vector;
+
+  constexpr explicit iterator(int __id) noexcept
+      : __id_(__id)
+  {}
+
+  int __id_ = 0;
+};
+
+_CCCL_NODISCARD inline constexpr device device_vector::operator[](_CUDA_VSTD::size_t __id_) const noexcept
+{
+  assert(__id_ < size());
+  return device(static_cast<int>(__id_));
+}
```

Comment:
Is there any need to define this outline of the class definition?

Reply:
not a strong one. the nested `iterator` struct is large so i wanted to define it out of line. which requires that `begin()` and `end()` need to also be out of line. and then it seemed weird to me to have some members be in situ and others out of line, so i moved them all out of line.